packages/protocol/test/stability/exchange.ts

```diff
@@ -376,13 +387,549 @@ contract('Exchange', (accounts: string[]) => {
+        let oldTotalSupply: BigNumber
+        beforeEach(async () => {
+          oldTotalSupply = await stableToken.totalSupply()
+          oldReserveGoldBalance = await goldToken.balanceOf(mockReserve.address)
+          await goldToken.approve(exchange.address, goldTokenAmount, { from: user })
+          oldGoldBalance = await goldToken.balanceOf(user)
+        })
+
+        it(`should increase the user's stable balance`, async () => {
+          await fn(
+            goldTokenAmount,
+            expectedStableBalance.integerValue(BigNumber.ROUND_FLOOR),
+            true,
+            {
+              from: user,
+            }
+          )
+          const newStableBalance = await stableToken.balanceOf(user)
+          assertEqualBN(newStableBalance, expectedStableBalance)
+        })
+
+        it(`should decrease the user's gold balance`, async () => {
+          await fn(
+            goldTokenAmount,
+            expectedStableBalance.integerValue(BigNumber.ROUND_FLOOR),
+            true,
+            {
+              from: user,
+            }
+          )
+          const actualGoldBalance = await goldToken.balanceOf(user)
+          let expectedGoldBalance = oldGoldBalance.minus(goldTokenAmount)
+          const block = await web3.eth.getBlock('latest')
+          if (isSameAddress(block.miner, user)) {
+            const blockReward = new BigNumber(2).times(new BigNumber(10).pow(decimals))
+            expectedGoldBalance = expectedGoldBalance.plus(blockReward)
+          }
+          assertEqualBN(actualGoldBalance, expectedGoldBalance)
+        })
+
+        it(`should remove the user's allowance`, async () => {
+          await fn(
+            goldTokenAmount,
+            expectedStableBalance.integerValue(BigNumber.ROUND_FLOOR),
+            true,
+            {
+              from: user,
+            }
+          )
+          const allowance = await goldToken.allowance(user, exchange.address)
+          assert.isTrue(allowance.isZero())
+        })
+
+        it(`should increase the Reserve's balance`, async () => {
+          await fn(
+            goldTokenAmount,
+            expectedStableBalance.integerValue(BigNumber.ROUND_FLOOR),
+            true,
+            {
+              from: user,
+            }
+          )
+          const newReserveGoldBalance = await goldToken.balanceOf(mockReserve.address)
+          assert.isTrue(newReserveGoldBalance.eq(oldReserveGoldBalance.plus(goldTokenAmount)))
+        })
+
+        it('should increase the total StableToken supply', async () => {
+          await fn(
+            goldTokenAmount,
+            expectedStableBalance.integerValue(BigNumber.ROUND_FLOOR),
+            true,
+            {
+              from: user,
+            }
+          )
+          const newTotalSupply = await stableToken.totalSupply()
+          assert.isTrue(newTotalSupply.eq(oldTotalSupply.plus(expectedStableBalance)))
+        })
+
+        it('should affect token supplies', async () => {
```

Comment:
Maybe supplies -> buckets, since supply can be confused for global floating supply?